File: Burning Tree - GFG/burning-tree.cpp
// { Driver Code Starts
//Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

struct Node {
    int data;
    Node *left;
    Node *right;

    Node(int val) {
        data = val;
        left = right = NULL;
    }
};


Node *buildTree(string str) {
    // Corner Case
    if (str.length() == 0 || str[0] == 'N')
        return NULL;

    // Creating vector of strings from input
    // string after spliting by space
    vector<string> ip;

    istringstream iss(str);
    for (string str; iss >> str;)
        ip.push_back(str);

    // Create the root of the tree
    Node *root = new Node(stoi(ip[0]));

    // Push the root to the queue
    queue<Node *> queue;
    queue.push(root);

    // Starting from the second element
    int i = 1;
    while (!queue.empty() && i < ip.size()) {

        // Get and remove the front of the queue
        Node *currNode = queue.front();
        queue.pop();

        // Get the current Node's value from the string
        string currVal = ip[i];

        // If the left child is not null
        if (currVal != "N") {

            // Create the left child for the current Node
            currNode->left = new Node(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->left);
        }

        // For the right child
        i++;
        if (i >= ip.size())
            break;
        currVal = ip[i];

        // If the right child is not null
        if (currVal != "N") {

            // Create the right child for the current Node
            currNode->right = new Node(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->right);
        }
        i++;
    }

    return root;
}


 // } Driver Code Ends
//User function Template for C++

class Solution {
  public:
    void burnTheTree(Node *root,int maxT,stack<Node *>&st,int &ans,Node* blocker){
        if(root==NULL || root==blocker) return;
        
        if(root->left!=NULL){
            burnTheTree(root->left,maxT+1,st,ans,blocker);
        }
        if(root->right!=NULL){
            burnTheTree(root->right,maxT+1,st,ans,blocker);
        }
        if(st.top() == root){
            st.pop();
            if(!st.empty()) burnTheTree(st.top(),maxT+1,st,ans,root);
        }
        ans = max(maxT,ans);
    }
    void rootToNodePath(Node *root,int target,vector<Node*> &v,stack<Node*>&s){
        if(root == NULL) return;
        v.push_back(root);
        if(root->data==target)
        {
            for(int i =0 ;i<v.size() ;i++) s.push(v[i]);
            return;
        }
        rootToNodePath(root->left,target,v,s);
        rootToNodePath(root->right,target,v,s);
        v.pop_back();
    }
    int minTime(Node* root, int target) 
    {
        // Your code goes here
        int ans = 0;
        stack<Node*> st;
        vector<Node*> v;
        rootToNodePath(root,target,v,st);
        Node* blocker = NULL;
        burnTheTree(st.top(),0,st,ans,blocker);
        return ans;
    }
};

// { Driver Code Starts.

int main() 
{
    int tc;
    scanf("%d ", &tc);
    while (tc--) 
    {    
        string treeString;
        getline(cin, treeString);
        // cout<<treeString<<"\n";
        int target;
        cin>>target;
        // cout<<target<<"\n";

        Node *root = buildTree(treeString);
        Solution obj;
        cout<<obj.minTime(root, target)<<"\n"; 

        cin.ignore();

    }


    return 0;
}
  // } Driver Code Ends

File: Burning Tree - GFG/burning-tree.java
// { Driver Code Starts
//Initial Template for Java


import java.util.LinkedList;
import java.util.Queue;
import java.io.*;
import java.util.*;

class Node {
	int data;
	Node left;
	Node right;

	Node(int data) {
		this.data = data;
		left = null;
		right = null;
	}
}

class GfG {

	static Node buildTree(String str) {

		if (str.length() == 0 || str.charAt(0) == 'N') {
			return null;
		}

		String ip[] = str.split(" ");
		// Create the root of the tree
		Node root = new Node(Integer.parseInt(ip[0]));
		// Push the root to the queue

		Queue<Node> queue = new LinkedList<>();

		queue.add(root);
		// Starting from the second element

		int i = 1;
		while (queue.size() > 0 && i < ip.length) {

			// Get and remove the front of the queue
			Node currNode = queue.peek();
			queue.remove();

			// Get the current node's value from the string
			String currVal = ip[i];

			// If the left child is not null
			if (!currVal.equals("N")) {

				// Create the left child for the current node
				currNode.left = new Node(Integer.parseInt(currVal));
				// Push it to the queue
				queue.add(currNode.left);
			}

			// For the right child
			i++;
			if (i >= ip.length)
				break;

			currVal = ip[i];

			// If the right child is not null
			if (!currVal.equals("N")) {

				// Create the right child for the current node
				currNode.right = new Node(Integer.parseInt(currVal));

				// Push it to the queue
				queue.add(currNode.right);
			}
			i++;
		}

		return root;
	}

	static void printInorder(Node root) {
		if (root == null)
			return;

		printInorder(root.left);
		System.out.print(root.data + " ");

		printInorder(root.right);
	}

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		int t = Integer.parseInt(br.readLine());

		while (t > 0) {
			String s = br.readLine();
			int target = Integer.parseInt(br.readLine());
			Node root = buildTree(s);

			Solution g = new Solution();
			System.out.println(g.minTime(root, target));
			t--;

		}
	}
}


// } Driver Code Ends


//User function Template for Java

class Solution
{
    /*class Node {
    	int data;
    	Node left;
    	Node right;
    
    	Node(int data) {
    		this.data = data;
    		left = null;
    		right = null;
    	}
    }*/
    
public static int minTime(Node root, int target) {
    // Your code goes here
    Map<Node, Node> map = new HashMap<>();
    Node[] targetNode = new Node[1];
    fill(root, target, map, targetNode);
    if(targetNode[0] == null) {
        return 0;
    }

    return findMinTime(targetNode[0], map);
}

private static void fill(Node root, int target, Map<Node, Node> map, Node[] targetNode) {
    if(root == null) {
        return;
    }
    if(root.data == target) {
        targetNode[0] = root;
    }
    if (root.left != null) {
        map.put(root.left, root);
        fill(root.left, target, map, targetNode);
    }
    if (root.right != null) {
        map.put(root.right, root);
        fill(root.right, target, map, targetNode);
    }
}

private static int findMinTime(Node targetNode, Map<Node, Node> map) {
    Map<Node, Boolean> visited = new HashMap<>();
    Queue<Node> q = new LinkedList<>();
    q.add(targetNode);
    visited.put(targetNode, true);

    int result = 0;
    while(!q.isEmpty()) {
        int size = q.size();
        boolean anyBurnt = false;
        for (int i = 0; i < size; i++) {
            Node n = q.remove();
            if(n.left != null && !visited.containsKey(n.left)) {
                q.add(n.left);
                visited.put(n.left, true);
                anyBurnt = true;
            }
            if(n.right != null && !visited.containsKey(n.right)) {
                q.add(n.right);
                visited.put(n.right, true);
                anyBurnt = true;
            }
            if(map.containsKey(n) && !visited.containsKey(map.get(n))) {
                q.add(map.get(n));
                visited.put(map.get(n), true);
                anyBurnt = true;
            }
        }
        if(anyBurnt) {
            result++;
        }
    }
    return result;
}

}